### PR TITLE
Feature/uar 1151 individual mo residential address displayed

### PR DESCRIPTION
**Pull request by bmcgreevyCH**
### JIRA link
https://companieshouse.atlassian.net/browse/UAR-1151

### Change description
Individual mo residential address displayed

### CODE REVIEW

File: src/utils/update/psc.to.beneficial.owner.type.mapper.ts [hunk 1]
```diff
@@ -5,6 +5,7 @@ import { BeneficialOwnerOther } from "../../model/beneficial.owner.other.model";
 import { NatureOfControlType, yesNoResponse } from "../../model/data.types.model";
 import { mapBOMOAddress, isSameAddress, mapDateOfBirth, mapSelfLink, mapInputDate, splitNationalities, mapBOIndividualName } from "./mapper.utils";
 import { logger } from "../../utils/logger";
+import { BeneficialOwnerPrivateData } from "@companieshouse/api-sdk-node/dist/services/overseas-entities";
```
Comment (AndreKramer7654) [inline src/utils/update/psc.to.beneficial.owner.type.mapper.ts:8]: This is a bo change - happy to keep here.

File: test/utils/update/beneficial_owners_managing_officers_data_fetch.spec.ts [hunk 2]
```diff
@@ -43,5 +55,31 @@ describe("util beneficial owners managing officers data fetch", () => {
...
+    appData = { "transaction_id": undefined, "overseas_entity_id": undefined, "entity_number": "1234" };
+
+    mockGetCompanyOfficers.mockReturnValue(MOCK_GET_COMPANY_OFFICERS);
+
+    await retrieveManagingOfficers(req, appData);
+
+    expect(mockGetCompanyOfficers).toHaveBeenCalled();
+    expect(mockGetManagingOfficerPrivateData).not.toHaveBeenCalled();
+    expect(mockLoggerInfo).toHaveBeenCalled();
+
```
Comment (AndreKramer7654) [inline test/utils/update/beneficial_owners_managing_officers_data_fetch.spec.ts:71]: remove newline

File: test/utils/update/fetch.managing.officers.private.data.spec.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+const mockIsActiveFeature = isActiveFeature as jest.Mock;
+const mockGetManagingOfficersPrivateData = getManagingOfficersPrivateData as jest.Mock;
+const mockLoggerInfo = logger.info as jest.Mock;
+const mockLoggerError = logger.errorRequest as jest.Mock;
+
+describe("fetchManagingOfficersPrivateData", () => {
+  let appData: ApplicationData, req: Request;
+
+  beforeEach(() => {
+    appData = {
```
Comment (AndreKramer7654) [inline test/utils/update/fetch.managing.officers.private.data.spec.ts:22]: maybe make this a mock object too

**Review by AndreKramer7654 — APPROVED** [review]

File: test/utils/update/beneficial_owners_managing_officers_data_fetch.spec.ts [hunk 4]
```diff
@@ -43,5 +55,30 @@ describe("util beneficial owners managing officers data fetch", () => {
     retrieveBoAndMoData(req, appData);
     expect(mockGetCompanyPscService).not.toHaveBeenCalled();
     expect(mockGetCompanyOfficers).not.toHaveBeenCalled();
+    expect(mockGetManagingOfficerPrivateData).not.toHaveBeenCalled();
+  });
+
+  test("Should not call getManagingOfficerPrivateData when transactionId and overseasEntityId are undefined", async () => {
```
Comment (n-r-b) [inline test/utils/update/beneficial_owners_managing_officers_data_fetch.spec.ts:61]: I'm assuming these tests are no longer needed for this service?

File: src/utils/update/fetch.managing.officers.private.data.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        mapManaginOfficersPrivateData(moPrivateData, appData);
+        // Note: saved to persistent session when appData.entity.email is fetched.
+      }
+    } catch (error) {
+      logger.errorRequest(req, `Private Managing Officer details could not be retrieved for overseas entity ${appData.entity_number}`);
+    }
+  }
+};
+
+const mapManaginOfficersPrivateData = (moPrivateData, appData: ApplicationData) => {
```
Comment (n-r-b) [inline src/utils/update/fetch.managing.officers.private.data.ts:34]: Please correct the typo

File: test/utils/update/fetch.managing.officers.private.data.spec.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    jest.clearAllMocks();
+  });
+
+  test("should map Managing Officers Private Data when data is available", async () => {
+    mockIsActiveFeature.mockReturnValue(false);
+    mockGetManagingOfficersPrivateData.mockResolvedValue(MOCK_MANAGING_OFFICERS_PRIVATE_DATA);
+
+    await fetchManagingOfficersPrivateData(appData, req);
+
+    expect(appData.update?.review_managing_officers_individual?.length).toEqual(1);
```
Comment (n-r-b) [inline test/utils/update/fetch.managing.officers.private.data.spec.ts:32]: We could also added the equivalent check for corporate MOs:

`expect(appData.update?.review_managing_officers_corporate?.length).toEqual(1);`

File: test/utils/update/fetch.managing.officers.private.data.spec.ts [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    if (appData.update && appData.update.review_managing_officers_individual) {
+      appData.update.review_managing_officers_individual[0].ch_reference = '';
+    }
+
+    mockIsActiveFeature.mockReturnValue(false);
+    mockGetManagingOfficersPrivateData.mockResolvedValue(MOCK_MANAGING_OFFICERS_PRIVATE_DATA);
+
+    await fetchManagingOfficersPrivateData(appData, req);
+    expect(mockGetManagingOfficersPrivateData).toHaveBeenCalled();
+    expect(appData.update?.review_managing_officers_individual?.[0].usual_residential_address).toBeUndefined;
```
Comment (n-r-b) [inline test/utils/update/fetch.managing.officers.private.data.spec.ts:116]:
```suggestion
    expect(appData.update?.review_managing_officers_individual?.[0].usual_residential_address).toBeUndefined();
```

File: test/utils/update/fetch.managing.officers.private.data.spec.ts [hunk 8]
```diff
@@ -0,0 +1,127 @@
+import { describe, expect, test, beforeEach, jest } from '@jest/globals';
+import { fetchManagingOfficersPrivateData } from "../../../src/utils/update/fetch.managing.officers.private.data";
+import { getManagingOfficersPrivateData } from "../../../src/service/private.overseas.entity.details";
+import { isActiveFeature } from "../../../src/utils/feature.flag";
+import { logger } from "../../../src/utils/logger";
+import { ApplicationData } from "../../../src/model";
+import { MOCK_MANAGING_OFFICERS_PRIVATE_DATA, MOCK_APP_DATA_MOS } from "../../../test/__mocks__/session.mock";
```
Comment (n-r-b) [inline test/utils/update/fetch.managing.officers.private.data.spec.ts:7]:
```suggestion
import { MOCK_MANAGING_OFFICERS_PRIVATE_DATA, MOCK_APP_DATA_MOS } from "../../__mocks__/session.mock";
```

File: test/utils/update/beneficial_owners_managing_officers_data_fetch.spec.ts [hunk 9]
```diff
@@ -1,30 +1,42 @@
...
 import { getCompanyOfficers } from "../../../src/service/company.managing.officer.service";
 import { MOCK_GET_COMPANY_PSC_ALL_BO_TYPES } from "../../__mocks__/get.company.psc.mock";
 import { MOCK_GET_COMPANY_OFFICERS } from '../../__mocks__/get.company.officers.mock';
+import { getManagingOfficersPrivateData } from '../../../src/service/private.overseas.entity.details';
+import { MOCK_MO_PRIVATE_DATA } from '../../__mocks__/get.managing.officer.private.data.mock';
+import { logger } from '../../../src/utils/logger';
 
 const mockGetCompanyPscService = getCompanyPsc as jest.Mock;
 const mockGetCompanyOfficers = getCompanyOfficers as jest.Mock;
+const mockGetManagingOfficerPrivateData = getManagingOfficersPrivateData as jest.Mock;
```
Comment (n-r-b) [inline test/utils/update/beneficial_owners_managing_officers_data_fetch.spec.ts:20]: The private data mocks can be removed

File: test/utils/update/fetch.managing.officers.private.data.spec.ts [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    if (appData.update && appData.update.review_managing_officers_individual) {
+      appData.update.review_managing_officers_individual[0].ch_reference = undefined;
+    }
+
+    mockIsActiveFeature.mockReturnValue(false);
+    mockGetManagingOfficersPrivateData.mockResolvedValue(MOCK_MANAGING_OFFICERS_PRIVATE_DATA);
+
+    await fetchManagingOfficersPrivateData(appData, req);
+    expect(mockGetManagingOfficersPrivateData).toHaveBeenCalled();
+    expect(appData.update?.review_managing_officers_individual?.[0].usual_residential_address).toBeFalsy();
```
Comment (n-r-b) [inline test/utils/update/fetch.managing.officers.private.data.spec.ts:116]: Does `toBeFalsy()` not expect the address to be boolean?

Comment (AndreKramer7654) [reply test/utils/update/fetch.managing.officers.private.data.spec.ts:116]: everything is truthy in JavaScript :-)
maybe test for undefined instead?

File: src/utils/update/beneficial_owners_managing_officers_data_fetch.ts [hunk 11]
```diff
@@ -58,21 +59,37 @@ export const retrieveBeneficialOwners = async (req: Request, appData: Applicatio
...
 export const retrieveManagingOfficers = async (req: Request, appData: ApplicationData) => {
+
   const companyOfficers = await getCompanyOfficers(req, appData[EntityNumberKey] as string);
-  if (companyOfficers) {
-    for (const officer of (companyOfficers.items || [])) {
-      logger.info("Loaded officer " + officer.officerRole);
-      if (officer.resignedOn === undefined) {
-        if (officer.officerRole === "managing-officer") {
-          const managingOfficer = mapToManagingOfficer(officer);
-          logger.info("Loaded Managing Officer " + managingOfficer.id + " is " + managingOfficer.first_name + ", " + managingOfficer.last_name);
```
Comment (AndreKramer7654) [inline src/utils/update/beneficial_owners_managing_officers_data_fetch.ts:68]: please remove name from logs

File: src/utils/update/beneficial_owners_managing_officers_data_fetch.ts [hunk 12]
```diff
@@ -58,21 +59,37 @@ export const retrieveBeneficialOwners = async (req: Request, appData: Applicatio
...
+
+const handleIndividualManagingOfficer = (officer: CompanyOfficer, appData: ApplicationData) => {
+  const managingOfficer = mapToManagingOfficer(officer);
+  logger.info(`Loaded Managing Officer ${managingOfficer.id} is ${managingOfficer.first_name}, ${managingOfficer.last_name}`);
+  appData.update?.review_managing_officers_individual?.push(managingOfficer);
+};
+
+const handleCorporateManagingOfficer = (officer: CompanyOfficer, appData: ApplicationData) => {
+  const managingOfficerCorporate = mapToManagingOfficerCorporate(officer);
+  logger.info(`Loaded Corporate Managing Officer ${managingOfficerCorporate.id} is ${managingOfficerCorporate.name}`);
```
Comment (AndreKramer7654) [inline src/utils/update/beneficial_owners_managing_officers_data_fetch.ts:93]: please remove name from logs

File: src/utils/update/beneficial_owners_managing_officers_data_fetch.ts (hunk 11 above)
Comment (bmcgreevyCH) [reply src/utils/update/beneficial_owners_managing_officers_data_fetch.ts:68]: This is deleted but I'll do it in the code that has it.

**Review by n-r-b — APPROVED** [review]

**Pull request merged**